src/js/components/Animate.js

```diff
@@ -0,0 +1,203 @@
+    this.enter(callback);
+  }
+
+  componentDidAppear () {
+    this.entered();
+  }
+
+  componentDidEnter () {
+    this.entered();
+  }
+
+  componentWillLeave (callback) {
+    this.leave(callback);
+  }
+
+  enter (callback) {
+    const {
+      enter: {
+        animation: enterAnimation,
+        delay
+      },
+      leave: {
+        animation: leaveAnimation
+      }
+    } = this.props;
+    const node = ReactDOM.findDOMNode(this);
+
+    const enterClass = enterAnimation;
+    const leaveClass = leaveAnimation || enterAnimation;
+    this.setState({enterClass});
+    this.setState({leaveClass});
+
+    if (node) {
+      node.style.transitionDuration = '';
+      node.classList.remove('animate', `${leaveClass}--leave`);
+      node.classList.add(`${enterClass}--enter`);
+      setTimeout(callback, delay);
+    }
+  }
+
+  entered () {
+    const { enter: { duration } } = this.props;
+    const node = ReactDOM.findDOMNode(this);
+    const enterClass = this.state.enterClass;
+    node.classList.remove(`${enterClass}--enter`);
+    if (duration) {
+      node.style.transitionDuration = `${duration}ms`;
+    }
+    node.classList.add('animate', `${enterClass}--enter-active`);
+    setTimeout(() => {
+      node.style.transitionDuration = '';
+      node.classList.remove('animate');
+    }, parseFloat(getComputedStyle(node).transitionDuration) * 1000);
+  }
+
+  leave (callback) {
+    const { leave: { duration, delay } } = this.props;
+    const node = ReactDOM.findDOMNode(this);
+
+    if (duration) {
+      node.style.transitionDuration = `${duration}ms`;
+    }
+
+    return setTimeout(() => {
+      node.classList.remove(`${this.state.enterClass}--enter-active`);
+      node.classList.add('animate', `${this.state.leaveClass}--leave`);
+      setTimeout(callback, parseFloat(getComputedStyle(node).transitionDuration) * 1000);
+    }, delay);
+  }
+
+  render () {
+    return this.props.children;
+  }
+};
+
+AnimateChild.defaultProps = {
+  enter: {},
+  leave: {}
+};
+
+
+class Animate extends Component {
+  constructor (props) {
+    super(props);
+    this.state = {
+      animationState: 'enter',
+      animation: props.enter.animation
+    };
+  }
+
+  componentWillReceiveProps (nextProps) {
+    const { visible, keep } = this.props;
+    const { animationState } = this.state;
+
+    if (keep && visible !== nextProps.visible) {
+      let state = '';
+      if (!nextProps.visible) {
+        state = (animationState === 'leave') ? 'enter' : 'leave';
+      }
+
+      const animateState = nextProps[state] || nextProps.enter;
+      this.setState({
+        animationState: state,
+        animation: state ? animateState.animation : ''
+      });
+
+      // Reset animation state back to enter after leave animation is finished
+      if (state === 'leave') {
+        const node = ReactDOM.findDOMNode(this);
+        clearTimeout(this.animationTimer);
+        this.animationTimer = setTimeout(() => {
+          this.setState({
```

Comment:
Is there any issue if the component continuing the animated component is unmounted before the timeout function is executed?